Improve precision of the vectorized trig functions

eiriktsarpalis opened the pull request:
Contributes to #98861.

eiriktsarpalis committed:
Improve precision of the vectorized trig functions

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.Sin.cs:
```diff
@@ -231,11 +231,17 @@ public static Vector128<double> Invoke(Vector128<double> x)
+                Vector128<double> f = uxMasked;
+                f = MultiplyAddEstimateOperator<double>.Invoke(dn, Vector128.Create(-double.Pi), f);
+                f = MultiplyAddEstimateOperator<double>.Invoke(dn, Vector128.Create(-Pi_Tail1), f);
+                f = MultiplyAddEstimateOperator<double>.Invoke(dn, Vector128.Create(-Pi_Tail2), f);
```
Were these changes to use MultiplyAddEstimateOperator opportunistic or necessary for precision? Interestingly I hadn't initially used this or FusedMultiplyAdd anywhere, and added it in because "why not", but your including it here suggests to me that maybe the precision in the original amd code was relying on gcc to automatically used fused-multiply add?

stephentoub replied:
Then should we be using FusedMultiplyAdd throughout rather than MultiplyAddEstimateOperator? The former is supposed to always use FMA, even if it's more expensive to do so, whereas the latter is free to fall back to non-FMA if it's faster to do so.

eiriktsarpalis replied:
Good point, I think we should do that although presumably we should be replacing every use of `MultiplyAddEstimateOperator` in these algorithms?

stephentoub replied:
>> Then should we be using FusedMultiplyAddOperator throughout rather than MultiplyAddEstimateOperator?

> although presumably we should be replacing every use of MultiplyAddEstimateOperator in these algorithms

Yes, that's what I meant by "throughout" :smile:

eiriktsarpalis replied:
I was wondering if we could do that in a follow-up, given that this change specifically fixes precision for the trig methods.

stephentoub replied:
Sure.

eiriktsarpalis commented:
It appears the relevant tests are failing in the mono_interpreter leg of CI. I wonder if this is the working of `MultiplyAddEstimateOperator` falling back to non-fused operators.

eiriktsarpalis committed:
Ensure all trig algorithms use FMA where required.

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.Cos.cs:
```diff
@@ -130,22 +130,22 @@ public static Vector128<float> Invoke(Vector128<float> x)
 
                 Vector128<float> almHuge = Vector128.Create(AlmHuge);
-                Vector128<float> dn = MultiplyAddEstimateOperator<float>.Invoke(uxMasked + Vector128.Create(float.Pi / 2), Vector128.Create(1 / float.Pi), almHuge);
+                Vector128<float> dn = FusedMultiplyAddOperator<float>.Invoke(uxMasked + Vector128.Create(float.Pi / 2), Vector128.Create(1 / float.Pi), almHuge);
```
As noted in response to the main discussion thread...

AMD's does not require FMA and would not generate it on downlevel hardware.

We likewise should not require it because it will lead to poor performance on any hardware that pre-dates 2013, as well as for R2R code and the default NAOT experience.

eiriktsarpalis replied:
What do you suggest? Are you saying we should only use FMA for hardware that provides vectorized support for it?

tannergooding replied:
Yes, which is identical to how AOCL libm behaves when running on downlevel hardware.

For an individual `(x * y) + z`, FMA can only produce a max 1 ULP difference (that is the raw bitwise value is 1 bit higher or lower).

For the entire set of 7 multiply-adds, we should still be well within the 16 digit precision range for fused vs non fused.

eiriktsarpalis commented:
Moving from `MultiplyAddEstimateOperator` to `FusedMultiplyAddOperator` fixed the failing tests in mono interpreter.

eiriktsarpalis committed:
Revert "Ensure all trig algorithms use FMA where required."

This reverts commit c97605cbd391bed244f578fcc66b7ea1d258232f.

eiriktsarpalis committed:
Use factories in memberdata methods for improved type safety and inference.

eiriktsarpalis committed:
Update tolerances for hardware that doesn't support FMA.

eiriktsarpalis commented:
After discussions with @tannergooding we've concluded that `MultiplyAddEstimate` is the best approach when porting the numeric algorithms since it more closely follows the semantics of the C code. I've reverted the relevant and commit and updated tests so that tolerances are increased in hardware that doesn't support FMA.

eiriktsarpalis committed:
Increase tolerance for trig functions in non-FMA hardware.

eiriktsarpalis commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.Cos.cs:
```diff
@@ -241,11 +241,14 @@ public static Vector128<double> Invoke(Vector128<double> x)
                 Vector128<double> almHuge = Vector128.Create(AlmHuge);
-                Vector128<double> dn = (uxMasked * Vector128.Create(1 / double.Pi)) + Vector128.Create(double.Pi / 2) + almHuge;
+                Vector128<double> half = Vector128.Create(0.5);
+                Vector128<double> dn = (uxMasked * Vector128.Create(1 / double.Pi)) + half + almHuge;
```
The particular expression didn't contribute to any loss of precision in my tests (IIRC it eventually rounds off to an integer) so I didn't update. We probably should though for the sake of consistency.
